### PR TITLE
[mlir] Add reshape propagation patterns for tensor.pad

### PR DESCRIPTION
**Pull request by Max191**
This PR adds fusion by collapsing and fusion by expansion patterns for `tensor.pad` ops in ElementwiseOpFusion. Pad ops can be expanded or collapsed as long as none of the padded dimensions will be expanded or collapsed.

### CODE REVIEW

File: mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp [hunk 1]
```diff
@@ -1702,6 +1760,80 @@ class FoldWithProducerReshapeOpByCollapsing
...
+      if (reInd.size() == 1)
+        continue;
+      if (llvm::any_of(reInd, [&](int64_t ind) {
+            return low[ind] != 0 || high[ind] != 0;
+          })) {
+        return failure();
+      }
+    }
+
+    SmallVector<OpFoldResult> newLow, newHigh;
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:1796]: Could we reuse some of the implementation. Propagating expand_shape up or collapse_shape down are essentially the same.

File: mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp [hunk 2]
```diff
@@ -1937,6 +2069,8 @@ void mlir::linalg::populateFoldReshapeOpsByExpansionPatterns(
     const ControlFusionFn &controlFoldingReshapes) {
   patterns.add<FoldReshapeWithGenericOpByExpansion>(patterns.getContext(),
                                                     controlFoldingReshapes);
+  patterns.add<FoldPadWithProducerReshapeOpByExpansion>(patterns.getContext(),
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:2072]: AFAICS both the patterns added here are "propagating by collapsing". You are moving the expand_shape down and collapse_shape up. In both cases the pad is happening on collapsed dimensions. So you should add both to the `populateFoldReshapeOpsByCollapsingPatterns`.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
The control function is not being used. Please fix that.

File: mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp (hunk 2 above)
Comment (Max191) [reply mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:2072]: `FoldPadWithProducerReshapeOpByExpansion` pushes the producer collapse_shape down, expanding the pad. `FoldPadWithProducerReshapeOpByCollapsing` pushes the producer expand_shape down, collapsing the pad. I think these are in the right places here.

File: mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp (hunk 1 above)
Comment (Max191) [reply mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:1796]: They are different enough that I think it is a little difficult to reuse much implementation. They are both propagating the reshapes down, just different reshapes.

**Comment by Max191** [comment]
> The control function is not being used. Please fix that.

Oh oops, I forgot to fix this. Fixing now

File: mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:2082]: Ok, I twisted myself in a knot when looking at tests. You are right.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**